src/core/src/FabricGeometry.cpp

```diff
@@ -290,72 +290,188 @@ void FabricGeometry::setGeometry(
-
-    faceVertexCounts.fill(faceVertexCountsFabric);
-    indices.fill(faceVertexIndicesFabric);
-    positions.fill(pointsFabric);
-
-    *extentFabric = localExtent;
-    *worldExtentFabric = worldExtent;
-    *localToEcefTransformFabric = UsdUtil::glmToUsdMatrix(localToEcefTransform);
-    *worldPositionFabric = worldPosition;
-    *worldOrientationFabric = worldOrientation;
-    *worldScaleFabric = worldScale;
-
-    FabricUtil::setTilesetId(_pathFabric, tilesetId);
-
-    if (_debugRandomColors) {
-        const auto r = glm::linearRand(0.0f, 1.0f);
-        const auto g = glm::linearRand(0.0f, 1.0f);
-        const auto b = glm::linearRand(0.0f, 1.0f);
-        displayColorFabric[0] = pxr::GfVec3f(r, g, b);
-    } else {
-        displayColorFabric[0] = DEFAULT_VERTEX_COLOR;
-    }
-
-    displayOpacityFabric[0] = DEFAULT_VERTEX_OPACITY;
-
-    if (hasTexcoords) {
-        const auto& texcoords = hasImagery ? imageryTexcoords : texcoords_0;
-
-        srw.setArrayAttributeSize(_pathFabric, FabricTokens::primvars_st, texcoords.size());
-
-        auto stFabric = srw.getArrayAttributeWr<glm::fvec2>(_pathFabric, FabricTokens::primvars_st);
-
-        texcoords.fill(stFabric);
-    }
-
-    if (hasNormals) {
-        srw.setArrayAttributeSize(_pathFabric, FabricTokens::primvars_normals, normals.size());
-
-        auto normalsFabric = srw.getArrayAttributeWr<glm::fvec3>(_pathFabric, FabricTokens::primvars_normals);
-
-        normals.fill(normalsFabric);
-    }
-
-    if (hasVertexColors) {
-        srw.setArrayAttributeSize(_pathFabric, FabricTokens::primvars_vertexColor, vertexColors.size());
-
+    if (primitive.mode == 0) { //if tile is a point cloud
+
+        const auto numVoxels = positions.size();
+        const float quadHalfSize = 1.5f;
+        srw.setArrayAttributeSize(_pathFabric, FabricTokens::points, static_cast<size_t>(numVoxels * 8));
+        srw.setArrayAttributeSize(_pathFabric, FabricTokens::faceVertexCounts, numVoxels * 2 * 6);
+        srw.setArrayAttributeSize(_pathFabric, FabricTokens::primvars_displayColor, numVoxels * 8);
+
+        srw.setArrayAttributeSize(_pathFabric, FabricTokens::faceVertexIndices, numVoxels * 6 * 2 * 3);
+
+        auto pointsFabric = srw.getArrayAttributeWr<pxr::GfVec3f>(_pathFabric, FabricTokens::points);
+        auto faceVertexCountsFabric = srw.getArrayAttributeWr<int>(_pathFabric, FabricTokens::faceVertexCounts);
+        auto faceVertexIndicesFabric = srw.getArrayAttributeWr<int>(_pathFabric, FabricTokens::faceVertexIndices);
+
+        std::vector<glm::fvec3> vertexColorsData(numVoxels);
+        gsl::span<glm::fvec3> vertexColorsSpan(vertexColorsData);
+        if (hasVertexColors) {
+            vertexColors.fill(vertexColorsSpan);
+            srw.setArrayAttributeSize(_pathFabric, FabricTokens::primvars_vertexColor, numVoxels * 8);
+        }
         auto vertexColorsFabric = srw.getArrayAttributeWr<glm::fvec3>(_pathFabric, FabricTokens::primvars_vertexColor);
```

Comment:
@lilleyse That's true, it's not great. I left it there because I think that accessor would be in the code base for 1-2 weeks. I think this CPU approach to creating the voxels is causing a perf hit. It's very easy to kernelize. I think what we'll have is `VertexColorsAccessor::data()` for consistency with accessing the data in a `span`. I also don't think our code-analysis tools check for functions with no references, so I'd need to keep a TODO to remove it. But it's not a huge deal either way. What do you think? Add it now and remove it later for the kernel, or swap it out for the kernel pretty soon?

Reply:
Alright, if CUDA is going to be generating geometry for us in the near future then it's ok as a vector.